target/qemu-m4/m4display.c

```diff
@@ -0,0 +1,217 @@
+#define CMD_DISPLAY_FRAME_BEGIN 0x05
+
+#define SCENE_BLACK  0x00
+#define SCENE_SPLASH 0x01
+#define SCENE_UPDATE 0x02
+#define SCENE_ERROR  0x03
+
+#define M4DISPLAY_WIDTH  180
+#define M4DISPLAY_HEIGHT 180
+
+static uint8_t framebuffer[M4DISPLAY_HEIGHT][M4DISPLAY_WIDTH];
+
+const char programming_header[] = "  Lattice\0iCEcube2 2014.08.26723\0Part: iCE40LP1K-CM36\0Date: Jan 30 2015 15:11:";
+
+void chip_select(bool val)
+{
+    if (val) {
+        gpio_set(GPIO(GPIO_PORT_G, 8), true);
+    } else {
+        gpio_set(GPIO(GPIO_PORT_G, 8), false);
+    }
+}
+
+void reset(bool val)
+{
+    if (val) {
+        gpio_set(GPIO(GPIO_PORT_G, 15), true);
+    } else {
+        gpio_set(GPIO(GPIO_PORT_G, 15), false);
+    }
+}
+
+void setup_pins(void)
+{
+
+    GPIO_InitTypeDef GPIO_InitStruct;
+
+    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOG, ENABLE);
+
+    GPIO_InitStruct.GPIO_Pin = GPIO_Pin_7 | GPIO_Pin_6 | GPIO_Pin_5;
+    GPIO_InitStruct.GPIO_Mode = GPIO_Mode_AF;
+    GPIO_InitStruct.GPIO_OType = GPIO_OType_PP;
+    GPIO_InitStruct.GPIO_Speed = GPIO_Speed_50MHz;
+    GPIO_InitStruct.GPIO_PuPd = GPIO_PuPd_NOPULL;
+    GPIO_Init(GPIOA, &GPIO_InitStruct);
+
+    // connect SPI6 pins to SPI alternate function
+    GPIO_PinAFConfig(GPIOA, GPIO_PinSource5, GPIO_AF_SPI6);
+    GPIO_PinAFConfig(GPIOA, GPIO_PinSource6, GPIO_AF_SPI6);
+    GPIO_PinAFConfig(GPIOA, GPIO_PinSource7, GPIO_AF_SPI6);
+
+    // enable clock for used IO pins
+    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_GPIOG, ENABLE);
+
+    /* Configure the chip select pin
+       in this case we will use PE7 */
+    GPIO_InitStruct.GPIO_Pin = GPIO_Pin_8;
+    GPIO_InitStruct.GPIO_Mode = GPIO_Mode_OUT;
+    GPIO_InitStruct.GPIO_OType = GPIO_OType_PP;
+    GPIO_InitStruct.GPIO_Speed = GPIO_Speed_50MHz;
+    GPIO_InitStruct.GPIO_PuPd = GPIO_PuPd_UP;
+    GPIO_Init(GPIOG, &GPIO_InitStruct);
+
+    GPIOE->BSRRL |= GPIO_Pin_7; // set PE7 high
+
+    // Setup display CS Pin
+    gpio_config(GPIO(GPIO_PORT_G, 8), GPIO_OUTPUT);
+
+    // Setup display reset pin
+    gpio_config(GPIO(GPIO_PORT_G, 15), GPIO_OUTPUT);
+
+    // enable peripheral clock
+    RCC_APB2PeriphClockCmd(RCC_APB2Periph_SPI6, ENABLE);
+}
+
+void init_display(void)
+{
+    // Setting up pins.
+    setup_pins();
+
+    SPI_InitTypeDef init_struct;
+
+    init_struct.SPI_Direction           = SPI_Direction_1Line_Tx;
+    init_struct.SPI_Mode                = SPI_Mode_Master;
+    init_struct.SPI_DataSize            = SPI_DataSize_8b;
+    init_struct.SPI_CPOL                = SPI_CPOL_Low;
+    init_struct.SPI_CPHA                = SPI_CPHA_1Edge;
+    init_struct.SPI_NSS                 = SPI_NSS_Hard;
+    init_struct.SPI_BaudRatePrescaler   = SPI_BaudRatePrescaler_8;
+    init_struct.SPI_FirstBit            = SPI_FirstBit_MSB;
+
+    SPI_Init(SPI6, &init_struct);
+
+    SPI_Cmd(SPI6, ENABLE); // enable SPI6
+
+    chip_select(true);
+    reset(true);
+    chip_select(false);
+
+    uint8_t draw_splash_cmds[] = { 0x04, 0x01 };
```

Comment:
can make these const, which should make the compiler not make a separate copy, though it likely matters not at all